### PR TITLE
[Trilinos] add monotonicity preserving solver

### PR DESCRIPTION
**Pull request by ddiezrod**
**Description**
Same as https://github.com/KratosMultiphysics/Kratos/pull/7409 but using Trilinos.
I checked the results and they are the same as in serial, but please take a look @RiccardoRossi as I took most of it from trilinos_block_builder_and_solver and I am not 100% whats going on here ;)

**Changelog**
-Added monotonicity preserving solver in TrilinosApp

### CODE REVIEW

File: applications/TrilinosApplication/external_includes/trilinos_monotonicity_preserving_solver.h
Comment (philbucher) [inline applications/TrilinosApplication/external_includes/trilinos_monotonicity_preserving_solver.h:61]:
```suggestion
    : public MonotonicityPreservingSolver<TSparseSpaceType, TDenseSpaceType,  TReordererType>
```

from what I see only `ProvideAdditionalData` is different, hence I think it should be possible to derive in this way

**Review by philbucher — COMMENTED** [review]
minor comment, rest looks good to me

please add the same test you added in serial in trilinos tests: https://github.com/KratosMultiphysics/Kratos/blob/master/applications/TrilinosApplication/tests/test_trilinos_linear_solvers.py

File: applications/TrilinosApplication/external_includes/trilinos_monotonicity_preserving_solver.h
Comment (ddiezrod) [reply applications/TrilinosApplication/external_includes/trilinos_monotonicity_preserving_solver.h:61]: Im afraid this does not work because then it tries to instantiate MonotonicityPreservingSolver with the trilinos space so you get errors like this:
```

/home/daniel/KratosDEV/KratosGitHub/applications/TrilinosApplication/custom_factories/trilinos_linear_solver_factory.cpp:105:1:   required from here
/home/daniel/KratosDEV/KratosGitHub/kratos/linear_solvers/monotonicity_preserving_solver.h:201:36: error: ‘Kratos::MonotonicityPreservingSolver<Kratos::TrilinosSpace<Epetra_FECrsMatrix, Epetra_FEVector>, Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> >, Kratos::Reorderer<Kratos::TrilinosSpace<Epetra_FECrsMatrix, Epetra_FEVector>, Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> > > >::SparseMatrixType {aka class Epetra_FECrsMatrix}’ has no member named ‘value_data’
         double *values_vector = rA.value_data().begin();
```

Comment (philbucher) [reply applications/TrilinosApplication/external_includes/trilinos_monotonicity_preserving_solver.h:61]: you are right, I hadn't considered that...
Then I guess it is ok as is, just add the test then I can approve :)

**Comment by ddiezrod** [comment]
@philbucher I think I cannot create a test like those ones because here everything is done in "ProvideAdditionalData" which is a function that's called from the B&S, Im not sure how to solve this...

**Comment by RiccardoRossi** [comment]
maybe you can call the function ProvideAdditionalData within the test

**Comment by philbucher** [comment]
> @philbucher I think I cannot create a test like those ones because here everything is done in "ProvideAdditionalData" which is a function that's called from the B&S, Im not sure how to solve this...

but how did you test the serial version?

**Comment by ddiezrod** [comment]
> > @philbucher I think I cannot create a test like those ones because here everything is done in "ProvideAdditionalData" which is a function that's called from the B&S, Im not sure how to solve this...
> 
> but how did you test the serial version?

Its here:
https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/cpp_tests/linear_solvers/test_monotonicity_solver.cpp

I called that function in the cpp test. So what I did is to create all the matrices manually, I am not sure if I could try that in mpi : /

**Comment by philbucher** [comment]
I see, I thought you were still using `mm` files that contain the matrix

So it comes down to creating the Epetra stuffs by hand, which is certainly an annoying/difficult task

@KratosMultiphysics/mpi this is sth that we should probably implement "globally" at some point, in order to be able to test things like this

@ddiezrod I leave it up to you to do it, but I will not force it, since writing the test will probably require 10x the time you needed for the implementation :)
It is not impossible, but requires more effort (assembling the maps & graph for the matrix etc)

=> for me can be done in the future

**Review by philbucher — APPROVED** [review]
hopefully we have some facilities in the future to simplify the creation of distributed tests
probably #6623 will be useful, then we are also independent of what is used in the background (trilinos or our own stuff)

**Comment by ddiezrod** [comment]
Thanks @philbucher I'll keep in mind that a test is missing. If it makes you feel better a small simulation will be included in the altair QA using this solver ;)

**Comment by ddiezrod** [comment]
Is this also ok for you @RiccardoRossi ?

**Comment by philbucher** [comment]
(just a thought, you could write the mm file from your cpp test and then use it in the trilinos tests, like the tests for the other linear solvers work)

**Comment by ddiezrod** [comment]
> (just a thought, you could write the mm file from your cpp test and then use it in the trilinos tests, like the tests for the other linear solvers work)

Yep, that was the original idea, but the problem is that ProvideAdditionalData has to be called. Maybe if we expose this function to python I could make it work creating also a small mdpa to be read in mpi

**Comment by ddiezrod** [comment]
Can I merge this and add the test when the otther PR is ready? Im a little in a hurry (as usual)

**Comment by philbucher** [comment]
it is approved so ...

**Comment by ddiezrod** [comment]
> it is approved so ...

Ok thanks, just to be sure.

**Pull request merged**
